### PR TITLE
arch/makefile: preprocess link script to make configure more flexibly 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arch/makefile: preprocess link script to make configure more flexibly

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm/src/Makefile [hunk 1]
```diff
@@ -165,7 +166,10 @@ define LINK_ALLSYMS
 	$(Q) $(call DELFILE, allsyms.tmp allsyms$(OBJEXT))
 endef
 
-nuttx$(EXEEXT): $(HEAD_OBJ) board$(DELIM)libboard$(LIBEXT) $(ARCHSCRIPT)
+$(addsuffix .tmp,$(ARCHSCRIPT)): $(ARCHSCRIPT)
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:169]:
```suggestion
$(ARCHSCRIPT).tmp: $(ARCHSCRIPT)
```

File: arch/arm/src/Makefile [hunk 2]
```diff
@@ -165,7 +166,10 @@ define LINK_ALLSYMS
...
+	$(call PREPROCESS, $(patsubst %.tmp,%,$@), $@)
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:170]:
```suggestion
	$(call PREPROCESS, $<, $@)
```

File: arch/arm/src/Makefile [hunk 3]
```diff
@@ -184,6 +188,7 @@ ifneq ($(CONFIG_WINDOWS_NATIVE),y)
 	grep -v '\(compiled\)\|\(\$(OBJEXT)$$\)\|\( [aUw] \)\|\(\.\.ng$$\)\|\(LASH[RL]DI\)' | \
 	sort > $(TOPDIR)$(DELIM)System.map
 endif
+	$(Q) $(call DELFILE, $(addsuffix .tmp,$(ARCHSCRIPT)))
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:191]:
```suggestion
	$(Q) $(call DELFILE, $(ARCHSCRIPT).tmp)
```

File: arch/arm/src/Makefile [hunk 4]
```diff
@@ -220,6 +225,7 @@ clean:
 ifeq ($(BOARDMAKE),y)
 	$(Q) $(MAKE) -C board clean
 endif
+	$(call DELFILE, $(addsuffix .tmp,$(ARCHSCRIPT)))
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:228]:
```suggestion
	$(call DELFILE,$(ARCHSCRIPT).tmp)
```

File: arch/arm/src/Makefile [hunk 5]
```diff
@@ -165,7 +166,10 @@ define LINK_ALLSYMS
 	$(Q) $(call DELFILE, allsyms.tmp allsyms$(OBJEXT))
 endef
 
-nuttx$(EXEEXT): $(HEAD_OBJ) board$(DELIM)libboard$(LIBEXT) $(ARCHSCRIPT)
+$(addsuffix .tmp,$(ARCHSCRIPT)): $(ARCHSCRIPT)
+	$(call PREPROCESS, $(patsubst %.tmp,%,$@), $@)
+
+nuttx$(EXEEXT): $(HEAD_OBJ) board$(DELIM)libboard$(LIBEXT) $(addsuffix .tmp,$(ARCHSCRIPT))
```
Comment (xiaoxiang781216) [inline arch/arm/src/Makefile:172]:
```suggestion
nuttx$(EXEEXT): $(HEAD_OBJ) board$(DELIM)libboard$(LIBEXT) $(ARCHSCRIPT).tmp
```

File: arch/arm/src/Makefile (hunk 1 above)
Comment (anchao) [reply arch/arm/src/Makefile:169]: $(ARCHSCRIPT) will contain multiple targets

File: arch/arm/src/Makefile (hunk 5 above)
Comment (anchao) [reply arch/arm/src/Makefile:172]: $(ARCHSCRIPT)  will contain multiple targets

File: arch/arm/src/Makefile (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/arm/src/Makefile:169]: If so, we need use % rule here like .c->.o

Comment (anchao) [reply arch/arm/src/Makefile:169]: ARCHSCRIPT do not have uniform naming rules and may exist in different directories, so the % rules may not cover for all cases

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**